libnmstate/state.py

```diff
@@ -276,64 +276,41 @@ def merge_interfaces(self, other_state):
-            Interface.IPV6, self, other_state)
-        self_iface_route_sets = defaultdict(set)
-        absent_route_sets = set()
-        other_iface_route_sets = defaultdict(set)
-
-        for route in self._config_routes:
-            if route.get(Route.STATE) == Route.STATE_ABSENT:
-                absent_route_sets.add(RouteEntry(route))
-
-        for route in other_state._config_routes:
-            # The other_state(running state) does not have any absent
-            # route entry and all have NEXT_HOP_INTERFACE
-            iface_name = route[Route.NEXT_HOP_INTERFACE]
-            other_iface_route_sets[iface_name].add(RouteEntry(route))
+        other_routes = set()
+        for ifname, routes in six.viewitems(other_state.config_iface_routes):
+            if self.config_iface_routes.get(ifname):
+                other_routes |= set(routes)
+
+        self_routes = {
+            route
+            for routes in six.viewvalues(self.config_iface_routes)
+            for route in routes if not route.absent
+        }
 
-        for route in self._config_routes:
-            if route.get(Route.STATE) != Route.STATE_ABSENT:
-                # The absent route is discarded, so the metadata.py don't need
-                # handle it.
-                if Route.NEXT_HOP_INTERFACE in route:
-                    self_iface_route_sets[route[Route.NEXT_HOP_INTERFACE]].add(
-                        RouteEntry(route))
-
-        _validate_routes(self_iface_route_sets,
-                         iface_enable_states,
-                         ipv4_enable_states,
-                         ipv6_enable_states)
-
-        # Absent route can only remove routes in current state.
-        _apply_absent_routes(absent_route_sets, other_iface_route_sets)
-
-        for iface_name, route_set in six.viewitems(other_iface_route_sets):
-            # The other_state(running state) does not have any absent
-            # route entry and all have NEXT_HOP_INTERFACE
-            for route_obj in route_set:
-                if _route_is_valid(route_obj,
-                                   iface_enable_states,
-                                   ipv4_enable_states,
-                                   ipv6_enable_states):
-                    self_iface_route_sets[iface_name].add(route_obj)
-
-        merged_routes = []
-        for iface_name, route_set in six.viewitems(self_iface_route_sets):
-            merged_routes.extend(
-                list(route_obj.to_dict() for route_obj in route_set))
-
-        self._config_routes = merged_routes
+        absent_routes = set()
+        for routes in six.viewvalues(self.config_iface_routes):
+            for absent_route in (r for r in routes if r.absent):
+                absent_routes |= {
+                    r for r in other_routes if absent_route.match(r)
+                }
+                absent_routes |= {
+                    r for r in self_routes if absent_route.match(r)
+                }
```

Comment:
The absent route should not remove routes in desire state, as user want to remove old routes and add new one.

Reply:
I will extract it to its own PR.